### DISCUSSION TITLE
Iterate ALL entities (getting EntityRef/EntityMut)

### DISCUSSION THREAD

**Discussion by Moulberry** [bevyengine/bevy#5496] [Q&A] [answered]
Disclaimer: I'm aware that you should be using the Query system for the majority of use cases, however I have a use-case where I need to get an EntityRef to every entity. If you need me to elaborate, let me know.

To my understanding, the simplest way to achieve the question posed in the title is as follows:
```rust
world.query::<Entity>().for_each(&world, |id| {
    let entity_ref = world.entity(id);
    // ...
});
```

And while this works, it smells bad to me. It feels semantically equivalent to the following:
```rust
let map = HashMap::new();
for key in map.keys() {
    let value = map.get(key);
}
```

Is there a more elegant solution to iterating every entity in bevy?

Cheers,
Moulberry

**Comment by alice-i-cecile** [comment, marked as answer]
This is the best solution currently. `World.entity` will be O(1) at least, but I expect the overhead to be relatively high compared to an optimized archetype-based solution.

Could you open a feature request issue for this?  Ideally we could have a `&EntityRef` or `&mut EntityMut` in queries, which would automatically imply exclusive world access 🤔

**Reply by Moulberry** [reply]
https://github.com/bevyengine/bevy/issues/5504
